databuilder/extractor/bigquery_watermark_extractor.py

```diff
@@ -0,0 +1,140 @@
+
+LOGGER = logging.getLogger(__name__)
+
+
+class BigQueryWatermarkExtractor(BaseBigQueryExtractor):
+
+    def init(self, conf):
+        # type: (ConfigTree) -> None
+        BaseBigQueryExtractor.init(self, conf)
+        self.iter = iter(self._iterate_over_tables())
+
+    def get_scope(self):
+        # type: () -> str
+        return 'extractor.bigquery_watermarks'
+
+    def _is_sharded_table(self, table_id):
+        suffix = table_id[-8:]
+        return suffix.isdigit()
+
+    def _retrieve_tables(self, dataset):
+        # type: () -> Any
+        sharded_table_watermarks = {}
+
+        for page in self._page_table_list_results(dataset):
+            if 'tables' not in page:
+                continue
+
+            for table in page['tables']:
+                tableRef = table['tableReference']
+                table_id = tableRef['tableId']
+
+                if self._is_sharded_table(table_id):
+                    suffix = table_id[-8:]
+                    prefix = table_id[:-8]
+
+                    if prefix in sharded_table_watermarks:
+                        sharded_table_watermarks[prefix]['low'] = min(sharded_table_watermarks[prefix]['low'], suffix)
+                        sharded_table_watermarks[prefix]['high'] = max(sharded_table_watermarks[prefix]['high'], suffix)
+                    else:
+                        sharded_table_watermarks[prefix] = {'high': suffix, 'low': suffix, 'table': table}
+                else:
+                    partitions = self._get_partitions(table, tableRef)
+                    if not partitions:
+                        continue
+                    low, high = self._get_partition_watermarks(table, tableRef, partitions)
+                    yield low
+                    yield high
+
+            for prefix, td in sharded_table_watermarks.items():
+                table = td['table']
+                tableRef = table['tableReference']
+
+                yield Watermark(
+                    datetime.datetime.fromtimestamp(float(table['creationTime']) / 1000).strftime('%Y-%m-%d %H:%M:%S'),
+                    'bigquery',
+                    tableRef['datasetId'],
+                    prefix,
+                    '__table__={partition_id}'.format(partition_id=td['low']),
+                    part_type="low_watermark",
+                    cluster=tableRef['projectId']
+                )
+
+                yield Watermark(
+                    datetime.datetime.fromtimestamp(float(table['creationTime']) / 1000).strftime('%Y-%m-%d %H:%M:%S'),
+                    'bigquery',
+                    tableRef['datasetId'],
+                    prefix,
+                    '__table__={partition_id}'.format(partition_id=td['high']),
+                    part_type="high_watermark",
+                    cluster=tableRef['projectId']
+                )
+
+    def _get_partitions(self, table, tableRef):
+        if 'timePartitioning' not in table:
```

Comment:
@feng-tao for your q. about how bigquery identifies partitioned vs. non-partitioned table, this is it.

Reply:
thx